File: RdFinance2/src/ca/mss/rd/finance/mortgage/impl/MortgageAmortization.java
package ca.mss.rd.finance.mortgage.impl;

import java.util.Iterator;
import java.util.Map;

import ca.mss.rd.finance.mortgage.AmortizationType;
import ca.mss.rd.finance.mortgage.PaymentFrequency;
import ca.mss.rd.util.UtilMisc;



public class MortgageAmortization {
	
	final static public String className = MortgageAmortization.class.getName();
	final static public long serialVersionUID = className.hashCode();

	final static public int DEFAULT_COLUMN_WIDTH = 11;

	final static public double MINIMUM_PAYMENT = 10.0;

	final static public int C_NOP = 						0;
	final static public int C_BALANCE_IN_DATE = 			1;
	final static public int C_BALANCE_IN = 					2;
	final static public int C_INTEREST = 					3;
	final static public int C_INTEREST_PRC = 				4;
	final static public int C_PRINCIPAL = 					5;
	final static public int C_PRINCIPAL_PRC = 				6;
	final static public int C_PAYMENT_DATE = 				7;
	final static public int C_PAYMENT = 					8;
	final static public int C_PAYMENT_MO = 					9;
	
	final static public int C_PAYMENT_EXTRA = 				10;
	final static public int C_PAYMENT_EXTRA_MO = 			11;
	final static public int C_PAYMENT_EXTRA_YEAR = 			12;
	
	final static public int C_PAYMENT_EXTRA_PRC = 			13;
	final static public int C_PAYMENT_EXTRA_PRC_MO = 		14;
	final static public int C_PAYMENT_EXTRA_PRC_YEAR = 		15;
	
	final static public int C_FULL_PRINCIPAL = 				16;
	final static public int C_FULL_PRINCIPAL_PRC = 			17;
	final static public int C_PAYMENT_FULL = 				18;
	
	final static public int C_TOTAL_FULL_PAYMENT = 			19;
	final static public int C_TOTAL_EXTRA_PAYMENT = 		20;
	final static public int C_TOTAL_EXTRA_PRC = 			21;
	final static public int C_TOTAL_PAYMENT = 				22;
	final static public int C_TOTAL_INTEREST = 				23;
	final static public int C_TOTAL_INTEREST_PRC = 			24;
	
	final static public int C_TOTAL_PRINCIPAL = 			25;
	final static public int C_TOTAL_PRINCIPAL_PRC = 		26;
	
	final static public int C_BALANCE_OUT = 				27;
	final static public int C_BALANCE_OUT_DATE = 			28;
	final static public int C_BALANCE_OUT_YEAR = 			29;
	final static public int C_BALANCE_OUT_MONTH = 			30;
	final static public int C_BALANCE_OUT_MONTH_YEAR = 		31;
	final static public int C_BALANCE_OUT_MONTH_DAY = 		32;
	final static public int C_BALANCE_OUT_QUOTER = 			33;
	final static public int C_BALANCE_TERM = 				34;

	final static public int C_PAY_IN_MO = 					35;
	final static public int C_PAYMENT_DATE_PREV =			36;
	final static public int C_PAYMENT_DATE_NEXT =			37;

	final static public String[] defTitle = new String[]{
		"NOP",
		"InDay",
		"InBalance",
		"Interest",
		"Intrst,%",
		"Principal",
		"Prncpl,%",
		"Payday",
		"Payment",
		"PaymentMo",
		
		"ExtraPmnt",
		"ExtraMo",
		"ExtraYear",
		
		"Extra,%",
		"ExtraMo,%",
		"ExtraYe,%",
		
		"FullPrncpl",
		"FullPrncpl,%",
		"FullPmnt",
		
		"TtlFullPmnt",
		"TtlExtraPmnt",
		"TtlExtra,%",
		"TtlPayment",
		"TtlInterest",
		"TtlIntrst,%",
		
		"TtlPrincipal",
		"TtlPrncpl,%",
		
		"OutBalance",
		"OutDate",
		"OutYear",
		"OutMonth",
		"OutMonthYear",
		"OutMonthDay",
		"OutQuoter",
		"Term",
		"Pay#",
		"Prev",
		"Next"
		};

	final static public Map<String,String> title = UtilMisc.toMap(
			"NOP", "no",
			"InDay", "Start",
			"InBalance", "Balance",
			"Interest", "Interest",
			"Intrst,%", "Int,%",
			"Principal", "Principal",
			"Prncpl,%", "Pri,%",
			"Payday", "Date",
			"Payment", "Pay",
			"PaymentMo", "PayMo",

			"ExtraPmnt", "Extra",
			"ExtraMo", "ExtrMo",
			"ExtraYear", "ExtrYr",
			
			"Extra,%", "Ex,%",
			"ExtraMo,%", "ExM,%",
			"ExtraYe,%", "ExY,%",
			
			"FullPrncpl", "Principal",
			"FullPrncpl,%", "Pri,%",
			"FullPmnt", "Pay",
			
			"TtlFullPmnt", "Payment",
			"TtlExtraPmnt", "ExtraTTL",
			"TtlExtra,%", "ExTTL,%",
			"TtlPayment", "PaymentTTL",
			"TtlInterest", "InterestTTL",
			"TtlIntrst,%", "IntTTL,%",
			"TtlPrincipal", "PrincipalTTL",
			"TtlPrncpl,%", "PriTTL,%",
			
			"OutBalance", "Balance",
			"OutDate", "Date",
			"OutYear", "Year",
			"OutMonth", "Month",
			"OutMonthYear", "Date",
			"OutMonthDay", "Date",
			"OutQuoter", "Quoter",
			"Term", "Term",
			"Pay#", "P#",
			"Prev", "Prev",
			"Next", "Next"
	);
			
	final static public boolean[] isVisible = new boolean[defTitle.length];
	final static public int[] columnNumber = new int[defTitle.length];

	final public PaymentFrequency paymentFrequency;
	final public int termNumberOfPayments, totalNumberOfPayments; 
	final public MortgageContext context;
	
	public MortgageAmortization(MortgageContext context, PaymentFrequency paymentFrequency) {
		this.context = context;
		this.paymentFrequency = paymentFrequency;
		this.totalNumberOfPayments = context.duration.getPaymentsNo(paymentFrequency);
		this.termNumberOfPayments = MortgageDuration.getPaymentsNo(context.termYears, 0, paymentFrequency);
	}

	final static public int getColumnWidth(int order){
		switch( order ){
		case C_NOP:
		case C_BALANCE_TERM:
			return 3;
		case C_BALANCE_IN_DATE:
		case C_BALANCE_OUT_DATE:
		case C_PAYMENT_DATE:
			return 9;
		case C_BALANCE_OUT_YEAR:
			return 5;
		case C_BALANCE_OUT_MONTH:
			return 4;
		case C_BALANCE_OUT_MONTH_YEAR:
		case C_BALANCE_OUT_MONTH_DAY:
		case C_BALANCE_OUT_QUOTER:
			return 6;
		default:
			return DEFAULT_COLUMN_WIDTH;
		}
	}

	final static public int getColumnWidth(){
		return DEFAULT_COLUMN_WIDTH;
	}

	final static public String getColumnTitle(int order){
		return title.get(defTitle[order]);
	}
			
	final public Iterator<String[]> getIteratorStar(AmortizationType at){
		return getIteratorStar(at, 0, 0);
	}

	final public Iterator<String[]> getIteratorStar(AmortizationType at, int year, int term){
		return new MortgageAmortizationIteratorStar(this, at, year, term);
	}

	final public Iterator<MortgageAmortizationRow> getIterator(AmortizationType at, int year, int term){
		return new MortgageAmortizationIterator(this, at, year, term);
	}

	final public Iterator<MortgageAmortizationRow> getIterator(AmortizationType at){
		return getIterator(at, 0, 0);
	}

	static final public void setVisible(int index, boolean b){
		MortgageAmortization.isVisible[index] = b;
	}

	static final public void setColumnNumber(int index, int n){
		MortgageAmortization.columnNumber[index] = n;
		MortgageAmortization.isVisible[index] = true;
	}

	static final public boolean isVisible(int index){
		return MortgageAmortization.isVisible[index];
	}
	
	static final public String columnLabels(int[] column){
		String labels = "";
		
		for(int i=0; i<column.length; i++){
			if( i > 0 )
				labels += ",";
			labels += getColumnTitle(column[i]);
		}
		
		return labels;
	}
	

}

File: RdFinanceBD/src.cpp/MortgageAmortization.cpp
#include "MortgageAmortization.h"
#include "MortgageAmortizationIteratorStar.h"
#include "MortgageAmortizationIterator.h"

namespace ca
{
	namespace mss
	{
		namespace rd
		{
			namespace mortgage
			{
				namespace impl
				{
					using ca::mss::rd::mortgage::AmortizationType;
					using ca::mss::rd::mortgage::PaymentFrequency;
					using ca::mss::rd::util::UtilMisc;
const std::wstring MortgageAmortization::className = MortgageAmortization::typeid::getName();
const java::math::BigDecimal MortgageAmortization::MINIMUM_PAYMENT = java::math::BigDecimal(L"10.0");
const std::wstring MortgageAmortization::defTitle[38] = {L"NOP", L"InDay", L"InBalance", L"Interest", L"Intrst,%", L"Principal", L"Prncpl,%", L"Payday", L"Payment", L"PaymentMo", L"ExtraPmnt", L"ExtraMo", L"ExtraYear", L"Extra,%", L"ExtraMo,%", L"ExtraYe,%", L"FullPrncpl", L"FullPrncpl,%", L"FullPmnt", L"TtlFullPmnt", L"TtlExtraPmnt", L"TtlExtra,%", L"TtlPayment", L"TtlInterest", L"TtlIntrst,%", L"TtlPrincipal", L"TtlPrncpl,%", L"OutBalance", L"OutDate", L"OutYear", L"OutMonth", L"OutMonthYear", L"OutMonthDay", L"OutQuoter", L"Term", L"Pay#", L"Prev", L"Next"};
java::util::Map<std::wstring, std::wstring> *const MortgageAmortization::title = ca::mss::rd::util::UtilMisc::toMap(L"NOP", L"no", L"InDay", L"Start", L"InBalance", L"Balance", L"Interest", L"Interest", L"Intrst,%", L"Int,%", L"Principal", L"Principal", L"Prncpl,%", L"Pri,%", L"Payday", L"Date", L"Payment", L"Pay", L"PaymentMo", L"PayMo", L"ExtraPmnt", L"Extra", L"ExtraMo", L"ExtrMo", L"ExtraYear", L"ExtrYr", L"Extra,%", L"Ex,%", L"ExtraMo,%", L"ExM,%", L"ExtraYe,%", L"ExY,%", L"FullPrncpl", L"Principal", L"FullPrncpl,%", L"Pri,%", L"FullPmnt", L"Pay", L"TtlFullPmnt", L"Payment", L"TtlExtraPmnt", L"ExtraTTL", L"TtlExtra,%", L"ExTTL,%", L"TtlPayment", L"PaymentTTL", L"TtlInterest", L"InterestTTL", L"TtlIntrst,%", L"IntTTL,%", L"TtlPrincipal", L"PrincipalTTL", L"TtlPrncpl,%", L"PriTTL,%", L"OutBalance", L"Balance", L"OutDate", L"Date", L"OutYear", L"Year", L"OutMonth", L"Month", L"OutMonthYear", L"Date", L"OutMonthDay", L"Date", L"OutQuoter", L"Quoter", L"Term", L"Term", L"Pay#", L"P#", L"Prev", L"Prev", L"Next", L"Next");

					MortgageAmortization::MortgageAmortization(MortgageContext *context, PaymentFrequency *paymentFrequency) : paymentFrequency(paymentFrequency), termNumberOfPayments(MortgageDuration::getPaymentsNo(context->termYears, 0, paymentFrequency)), totalNumberOfPayments(context->duration->getPaymentsNo(paymentFrequency)), context(context)
					{
					}

					int MortgageAmortization::getColumnWidth(int order)
					{
						switch (order)
						{
						case C_NOP:
						case C_BALANCE_TERM:
							return 3;
						case C_BALANCE_IN_DATE:
						case C_BALANCE_OUT_DATE:
						case C_PAYMENT_DATE:
							return 9;
						case C_BALANCE_OUT_YEAR:
							return 5;
						case C_BALANCE_OUT_MONTH:
							return 4;
						case C_BALANCE_OUT_MONTH_YEAR:
						case C_BALANCE_OUT_MONTH_DAY:
						case C_BALANCE_OUT_QUOTER:
							return 6;
						default:
							return DEFAULT_COLUMN_WIDTH;
						}
					}

					int MortgageAmortization::getColumnWidth()
					{
						return DEFAULT_COLUMN_WIDTH;
					}

					std::wstring MortgageAmortization::getColumnTitle(int order)
					{
						return title->get(defTitle[order]);
					}

					Iterator<std::wstring[]> *MortgageAmortization::getIteratorStar(AmortizationType *at)
					{
						return getIteratorStar(at, 0, 0);
					}

					Iterator<std::wstring[]> *MortgageAmortization::getIteratorStar(AmortizationType *at, int year, int term)
					{
						return new MortgageAmortizationIteratorStar(this, at, year, term);
					}

					Iterator<MortgageAmortizationRow*> *MortgageAmortization::getIterator(AmortizationType *at, int year, int term)
					{
						return new MortgageAmortizationIterator(this, at, year, term);
					}

					Iterator<MortgageAmortizationRow*> *MortgageAmortization::getIterator(AmortizationType *at)
					{
						return getIterator(at, 0, 0);
					}

					void MortgageAmortization::setVisible(int index, bool b)
					{
						MortgageAmortization::isVisible_Renamed[index] = b;
					}

					void MortgageAmortization::setColumnNumber(int index, int n)
					{
						MortgageAmortization::columnNumber[index] = n;
						MortgageAmortization::isVisible_Renamed[index] = true;
					}

					bool MortgageAmortization::isVisible(int index)
					{
						return MortgageAmortization::isVisible_Renamed[index];
					}

					std::wstring MortgageAmortization::columnLabels(int column[])
					{
						std::wstring labels = L"";

						for (int i = 0; i < sizeof(column) / sizeof(column[0]); i++)
						{
							if (i > 0)
							{
								labels += std::wstring(L",");
							}
							labels += getColumnTitle(column[i]);
						}

						return labels;
					}
				}
			}
		}
	}
}
